### ISSUE TITLE
about padding token

### ISSUE THREAD

**Issue by shixishi** [bghira/SimpleTuner#1726]
For classifier-free diffusion guidance, embeddings need to be computed for empty text prompt. 
When the empty text prompt is input into the clip text encoder, the padding operation is first performed, adding a lot of padding tokens, and then calculating the embedding. 
So, for the clip text encoder, the computed embedding contains a lot of padding tokens, Why not do the "prompt_embeds * attention_mask.unsqueeze(-1).expand(prompt_embeds.shape)", same as in the def _encode_sd3_prompt_with_t5() function that removes noise caused by padding tokens.

**Comment by bghira** [comment]
that logic for t5 causes model collapse for sd3 since it wasnt trained with it. the same would happen for clip, the padding is meaningfully attended to by the model

**Comment by shixishi** [comment]
> that logic for t5 causes model collapse for sd3 since it wasnt trained with it. the same would happen for clip, the padding is meaningfully attended to by the model

Thank you for your reply. But I don't quite understand.
Can you explain it in more detail? Thank you

**Comment by bghira** [comment]
the bias inside the models is so heavily trained to expect padding tokens that when you eliminate them, it requires an absurd amount of time and compute density to alter it in a meaningful way without degradation.

the padding tokens are being attended-to and are transformed by each layer as it goes through the model. what ends up happening is that some signal leakage is likely stored in these positions, and the model has to learn how to handle the task without being lazy and leaking any of this signal or "mis-using" token positions by holding the information there.

a common tactic to address this is to **prepend** register positions to the inputs. these are zero-initialised and then typically discarded before the model outputs; the model learns that it can place "trash" (information about tokens that the model wants to disregard) there but can not rely on it being present. this behaves similarly to an attention sink, a common trick in LLMs, where extra tokens at the start of a sequence absorb a lot of "extra attention" and stabilise outputs, which could be because the variance of early predictions otherwise can be quite broad.

similarly, in LLMs, removing the early register tokens harms the model, and requires a concerted high-compute effort to realign the model.

relevant research i've referenced:
- https://arxiv.org/html/2501.06751v1
- https://arxiv.org/abs/2309.16588

**Comment by shixishi** [comment]
> the bias inside the models is so heavily trained to expect padding tokens that when you eliminate them, it requires an absurd amount of time and compute density to alter it in a meaningful way without degradation.
> 
> the padding tokens are being attended-to and are transformed by each layer as it goes through the model. what ends up happening is that some signal leakage is likely stored in these positions, and the model has to learn how to handle the task without being lazy and leaking any of this signal or "mis-using" token positions by holding the information there.
> 
> a common tactic to address this is to **prepend** register positions to the inputs. these are zero-initialised and then typically discarded before the model outputs; the model learns that it can place "trash" (information about tokens that the model wants to disregard) there but can not rely on it being present. this behaves similarly to an attention sink, a common trick in LLMs, where extra tokens at the start of a sequence absorb a lot of "extra attention" and stabilise outputs, which could be because the variance of early predictions otherwise can be quite broad.
> 
> similarly, in LLMs, removing the early register tokens harms the model, and requires a concerted high-compute effort to realign the model.
> 
> relevant research i've referenced:
> 
> * https://arxiv.org/html/2501.06751v1
> * https://arxiv.org/abs/2309.16588

So, if I fine-tune the Stable diffusion 3 model by using my own dataset, what is right to do is not mask the padding tokens for either the T5 encoder or the CLIP encoder?
Is that correct?
Thank you.

**Comment by bghira** [comment]
yes
